### DISCUSSION TITLE
Face training tab is buggy, can use quite some improvements

### DISCUSSION THREAD

**Discussion by borgqueenx** [blakeblackshear/frigate#19721] [Ask A Question]
There are several things to improve face training in frigate. i find the experience so far quite beta-like.

- First of all and most importantly, me and my wife who walk around the most in and around our house, have 13 cameras in total, but in the face training tab during many hours during the day, faces are not showing up for training. if i stand at a camera, it definitely recognises a face, but in the faces tab nothing new shows up for long periods of time. Frigate should take multiple face thumbnails when a face is detected over the course of some seconds in my opinion, so we have good material for training. i can walk by a camera right now, even face it from a very close distance, and my face wont show up in the face training tab.

- Secondly, when training a face, i often notice that the thumbnail is different from the snapshot. and then when training, i see the face in the snapshot appear as the trained image. this is often a face from a different angle..and since face training has some precise instructions how to achieve the best results, this is not great behavior.

- Thirdly and lastly, using frigate on mobile, after i train a face, the green confirmation block above in the UI that a face was trained goes over the back button. i now have to wait until this green block is gone, until i can go back and continue.

**Comment by inkeep-ai-assistant** [comment]
Your feedback covers three main points about face training in Frigate:

1. **Faces not consistently appearing in the training tab:**  
Frigate requires a detected person before it can attempt face recognition. If faces are not showing up in the training tab even when you are in front of a camera, possible reasons include insufficient camera resolution, detection thresholds being too high, or the system being overloaded (especially if using CPU for detection). Adjusting the `detection_threshold` and `min_area` settings for face recognition may help more faces appear for training. Face thumbnails are captured when a qualifying face is detected, but there is no documented batch capture over several seconds. The recommendation is to build your training set gradually as faces appear, focusing on high-quality, diverse images for best results[(1)](https://docs.frigate.video/configuration/face_recognition), [(2)](https://github.com/blakeblackshear/frigate/discussions/18611).

2. **Thumbnail/snapshot mismatch during training:**  
When training, each detected face is grouped and labeled with a name and a snapshot. The documentation recommends selecting clear, front-facing images for training and avoiding images that already score highly to prevent overfitting. The system does not guarantee that the displayed thumbnail will always match the precise moment or angle you expect, and not all detected images are suitable for training. Best practices suggest only training on clear, lower-scoring images and gradually expanding the dataset with diverse angles and conditions[(1)](https://docs.frigate.video/configuration/face_recognition).

3. **UI confirmation overlay blocking navigation on mobile:**  
There is no documentation or official note about the confirmation block overlapping the back button on mobile devices. This appears to be a UI/UX issue that is not addressed in the official documentation.

For suggestions and improvements to face training and the UI, you can open a feature request or provide feedback directly on the [Frigate GitHub repository](https://github.com/blakeblackshear/frigate)[(3)](https://github.com/blakeblackshear/frigate/issues/18666).

📚 **Sources:**
1. [Face Recognition](https://docs.frigate.video/configuration/face_recognition)
2. [Face Recogination - i cant train faces... or where is my mistake?](https://github.com/blakeblackshear/frigate/discussions/18611)
3. [Nice to have: Face thumbnail in dropdown next to name of trained faces.](https://github.com/blakeblackshear/frigate/issues/18666)

---

Have another question? Just tag [@inkeep](https://inkeep.com).

**Comment by borgqueenx** [comment]
Inkeep is not providing fixes, these are probably good improvements for 0.17.

**Comment by NickM-27** [comment]
Thanks for sharing, though to be clear a lot of the things you mention here are not what other users are experiencing. During the beta some users were very passionate about face recognition, so we are very confident that the feature works very well. I'll address some of your points specifically. 

> Frigate should take multiple face thumbnails when a face is detected over the course of some seconds in my opinion, so we have good material for training

It does, especially when you use frigate+ like you are. If you're not seeing this work consistently then that's a configuration problem on your side. Likely the filters are too strong. Since you use frigate+ you can use the debug live view to see when frigate detects a face and the area / score. 

> Secondly, when training a face, i often notice that the thumbnail is different from the snapshot. and then when training, i see the face in the snapshot appear as the trained image. this is often a face from a different angle..and since face training has some precise instructions how to achieve the best results, this is not great behavior.

If Frigate only uses the snapshot then you would only have one chance per person to recognize a face. Your previous point just said that you expect a lot of faces to be taken. Your first and second points contradict each other. The face that is used for training will always be the one shown. The snapshot has nothing to do with face training.

**Reply by NickM-27** [reply]
as an illustrative example of this, here is a person delivering a package. They were on the camera for a total of 6 seconds, and 8 images were saved

<img width="1252" height="250" alt="Screen Shot 2025-08-23 at 07 15 14 AM" src="https://github.com/user-attachments/assets/6bb6f625-d4fd-4af3-bf62-fbada6f11826" />
<img width="1245" height="256" alt="Screen Shot 2025-08-23 at 07 15 21 AM" src="https://github.com/user-attachments/assets/bb8ed30d-2c1f-4c4a-a6b1-3b6d315b6abd" />
<img width="1218" height="535" alt="Screen Shot 2025-08-23 at 07 15 31 AM" src="https://github.com/user-attachments/assets/dc28c1b7-9f91-4d15-97c2-056e61ec0233" />

**Comment by borgqueenx** [comment]
cant show a screenshot due to privacy, but i just got 6 faces from a single snapshot in the face training menu. i can select any face from that list of 6 (different positioned/angled) faces and train, but in the list with trained images for that person, the same face/image shows. so it takes the snapshot and uses that? or is the thumbnail used, but the snapshot shown as thumbnail in the faces of a specific person?
At the very least thats confusing. or it is buggy.

**Reply by NickM-27** [reply]
> so it takes the snapshot and uses that?

Like I said the snapshot is not used for anything face recognition related. It's simply there to give you context about what happened on the camera when a given person was detected. The faces are grouped by event (person object) in the train tab.

**Reply by borgqueenx** [reply]
> > so it takes the snapshot and uses that?
> 
> Like I said the snapshot is not used for anything face recognition related. It's simply there to give you context about what happened on the camera when a given person was detected. The faces are grouped by event (person object) in the train tab.

doesn't it make sense to use a image of the actual face? i used 2 photos from the set of 6, but they show as the same image. that is just weird. The reason you say about the snapshot giving context about what happened, does not make sense, since a snapshot isnt available in the menu about the specific person set of faces. just 2 times the same face shows, even though i trained 2 times a different face, i hope.

**Reply by hawkeye217** [reply]
During the lifecycle of a tracked object of a `person`, a `face` may be detected many times. The cropped portion is passed off to the face recognition pipeline and a cropped thumbnail is saved, which is what you see in the Train tab. Frigate does not save a full snapshot for every face it recognizes during the object lifecycle.

**Reply by NickM-27** [reply]
> The reason you say about the snapshot giving context about what happened, does not make sense

That was added specifically because users requested it so they could have context, so our user feedback has said strongly that this is the preferred behavior.

**Reply by borgqueenx** [reply]
weird...but ok if the majority prefers it. i see a face in the library i would expect thats a face that has been used for training. in my opinion that makes the most sense.

On another note, i also feel that the blur option (thats standard enabled) is not working great. here are some images i feel comfortable sharing because they are impossible to make something out of them. yet frigate tries a lot, and therefore obviously, fails. the percentage of certainty is also so high...
![Screenshot_20250825_121635](https://github.com/user-attachments/assets/86a3c0a5-f981-4279-ae9e-1b8ac5e6a9d0)

**Reply by NickM-27** [reply]
> i see a face in the library i would expect thats a face that has been used for training. in my opinion that makes the most sense.

It is not clear what you mean by that. If you choose a person's name in the drop down in the top left and see a face there, then yes that face is the exact image that was used for training. But those faces are not related to the snapshot feature in Frigate in any way, shape, or form. 

 > On another note, i also feel that the blur option (thats standard enabled) is not working great. here are some images i feel comfortable sharing because they are impossible to make something out of them. yet frigate tries a lot, and therefore obviously, fails. the percentage of certainty is also so high...

There are a handful of problems that I see here:
1. blur reduction will take off at most 6% from the score. If at its maximum it took off a high % then that means even slight blurring would reduce scores considerably which would make the feature much less usable
2. the fact that these images have such large area, even 10,000 px at one point and are so blurry point to the cameras having compression set too high or have bad motion handling. Perhaps there are some options for the camera that can improve this.
3. These blurry images scoring so highly in some cases likely means that there are images used for training which are too blurry or inconsistent. This is where you generally want to pick out the worst images, delete them, and then re-process these blurry images to make sure that the confidence goes down which would indicate that the bad images were removed.

**Reply by borgqueenx** [reply]
then either i am again confused about a thumbnail of a face, or its not working as it should.
let me try to clarify as much as possible.
If i select in the face library, a face that is up for training, that usually comes in rows of probably a single event... when selected/clicked, i get the face, and i get a snapshot. When i then click train and click the name. it says successfully trained. then, i select another face from the same row (i dont do that much) and train it for the same person as well. 
When i now go to the individual person's tab, who i just trained 2 images from, i see the face that was in the snapshot, twice. not the 2 faces i trained.
here are 3 examples i feel are no problem sharing them:
<img width="186" height="181" alt="image" src="https://github.com/user-attachments/assets/f76d5d97-bace-4b12-9619-ea1ea7777507" />
<img width="151" height="181" alt="image" src="https://github.com/user-attachments/assets/8fde7f3f-50d1-49c6-a537-3dd9fb46f5cd" />
<img width="118" height="174" alt="image" src="https://github.com/user-attachments/assets/a0248cff-2a7d-407c-9103-33eae0d22584" />

These are in the list of one individual face. Obviously, i never selected the above faces for training. they looked like actual faces in the tab where i can select faces for training.
But this is how it looks like in the face tab of this specific person.

Furthermore, it makes it impossible to decide which picture to remove from training that i think are bad.

Another example:
<img width="364" height="202" alt="image" src="https://github.com/user-attachments/assets/80abe5bb-a1d0-4e80-b527-e9b0aacb3ff3" />
These are 2 different faces i send for training. yet, because they were from the same row, they show as a single image. and the actual face is once again, completely different then the blurry unusable face here.

**Reply by NickM-27** [reply]
Okay, I see where the confusion is now. You should not be clicking on the images to train them, you need to use the train button below the image to train that image

![Screen_Shot_2025-03-27_at_16 00 19_PM](https://github.com/user-attachments/assets/5194a086-a852-43bd-bfa6-bde4bf1abdc4)

Clicking on the row of images just shows you the object details which is not specific to that face, it is just there to give context to when / what happened when those face images were recognized

**Reply by borgqueenx** [reply]
so should i remove blurry faces from the individual faces menu? Those are the images used for training? Because as i showed above, i have a lot of those. But those are not the faces i selected for training. It changes from face training menu, towards the individual person menu.

**Reply by NickM-27** [reply]
yes, you should definitely remove any image from the named libraries (all tabs except the `train` tab) that are not just of a face. These will definitely cause problems like the balloon issue you posted below. Once that is done, and the images highlighted in my screenshot are used, the exact image that you see is what will be used for training, and the results should improve.

**Reply by borgqueenx** [reply]
Thanks. But that also confirms that there is a bug somewhere, where you select a face to train, but it then trains a other face image that is different then what you selected. Not sure how to reproduce. I used facenet, the small model at the start, and now using the large model, didnt have the issue as of yet.

**Reply by NickM-27** [reply]
It is not a bug. You were using the `train face` button that is in the object details dialog. This will use the face from the snapshot to train. The primary function of this is so you can use the face from a snapshot that you view in the `Explore` tab to train as well. 

<img width="633" height="69" alt="Screen Shot 2025-08-25 at 08 10 26 AM" src="https://github.com/user-attachments/assets/ad2b5df8-9c04-4386-8fa6-c2e46f72888b" />

**Comment by borgqueenx** [comment]
its okay that frigate can mistake a new object for a human, thats what frigate+ training is for. or to simply ignore temporary objects, or making a mask.

But being 100% certain that this is -insert name-......is a bit extreme.
<img width="986" height="351" alt="Schermafbeelding 2025-08-25 143537" src="https://github.com/user-attachments/assets/48e6546b-1302-4315-8f7a-3c0fe9bfca1b" />

**Reply by NickM-27** [reply]
> But being 100% certain that this is -insert name-......is a bit extreme.

this comes down to your images that you choose for training

**Reply by borgqueenx** [reply]
> > But being 100% certain that this is -insert name-......is a bit extreme.
> 
> this comes down to your images that you choose for training

i am not choosing baloons to train faces. only faces, haha.

**Reply by NickM-27** [reply]
> i am not choosing baloons to train faces. only faces, haha.

sure, but when blurry or otherwise incorrect images that don't have face details clear are used in training, the model will return very basic face features in the embedding that will match with anything. This is a problem with the images used for training, if the face features are clear then this ballon would not match anything close to the features of a person's face and the score would be very low

**Comment by mayerwin** [comment]
This thread is basically why I ended up building a little companion for the curation side: https://github.com/mayerwin/frigate-better-face-recognition. It runs locally next to Frigate (CPU, no GPU) and uses a real face detector to drop the non-face crops, scores each crop for quality so blurry ones stay out of training, and gives you a quick way to label genuine faces or mark something "not a face" so look-alikes get filtered after that. Your labels still train Frigate's own recognizer. Built it to scratch my own itch, it's free and MIT.
